File: Day1/day1.dart
import 'dart:io';

const topElfCaloriesLength = 3;
final topThreeElfCalories = <int>[0, 0, 0];
final latestElfCalories = <int>[];

void main() async {
  final inputFile = File("input.txt");
  final inputLines = await inputFile.readAsLines();

  for (final line in inputLines) {
    if (line.isEmpty) {
      final totalCalori = latestElfCalories.reduce((value, sum) => value + sum);
      updateTopThree(totalCalori);
      latestElfCalories.clear();
      continue;
    }

    final calori = int.parse(line);
    latestElfCalories.add(calori);
  }

  print("[Part1] Most carried calories: ${topThreeElfCalories[0]}");
  final topThreeSum = topThreeElfCalories.reduce((value, sum) => value + sum);
  print("[Part2] Top three calories sum: $topThreeSum");
}

void updateTopThree(int newCalories) {
  for (int i = 0; i < topElfCaloriesLength; i++) {
    if (newCalories > topThreeElfCalories[i]) {
      final oldCalori = topThreeElfCalories[i];
      topThreeElfCalories[i] = newCalories;
      updateTopThree(oldCalori);
      break;
    }
  }
}

File: Day1/day1.py
file = open('input.txt', 'r')
input_lines = file.readlines()

latest_elf_calories = []
top_three_elf_calories = [0, 0, 0]
top_elf_calories_len = 3

def update_top_three(new_calories): 
    for i in range(top_elf_calories_len):
        if new_calories > top_three_elf_calories[i]:
            old_calori = top_three_elf_calories[i]
            top_three_elf_calories[i] = new_calories
            update_top_three(old_calori)
            break

for line in input_lines:
    line = line.strip()
    if len(line) == 0:
        total_calori = sum(latest_elf_calories)
        update_top_three(total_calori)
        latest_elf_calories.clear()
        continue

    latest_elf_calories.append(int(line))

print("[Part1] Most carried calories:", top_three_elf_calories[0])
top_three_sum = sum(top_three_elf_calories)
print("[Part2] Top three calories sum:", top_three_sum)

